File: 240126/정수 사각형 최장 증가 수열/lis-on-the-integer-grid.java
import java.util.*;

public class Main {
    public static final int INF = 100000000;

    public static int n;
    public static int[][] board;
    public static int[][] dp;

    public static int[] dr = new int[]{-1, 1, 0, 0};
    public static int[] dc = new int[]{0, 0, -1, 1};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        n = sc.nextInt();
        board = new int[n][n];
        dp = new int[n][n];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[i][j] = sc.nextInt();
                dp[i][j] = -1;
            }
        }

        int maxRet = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                maxRet = Math.max(maxRet, dpf(i, j));
            }
        }

        System.out.println(maxRet);

    }

    public static boolean inRange(int r, int c){
        return 0 <= r && r < n && 0 <= c && c < n;
    }

    public static int dpf(int r, int c) {
        if (dp[r][c] != -1) {
            return dp[r][c];
        }

        boolean isStart = true;
        int ret = -INF;
        for (int dir = 0; dir < 4; dir++) {
            int pre_r = r + dr[dir];
            int pre_c = c + dc[dir];
            if (inRange(pre_r, pre_c) && board[r][c] > board[pre_r][pre_c]) {
                isStart = false;
                ret = Math.max(ret, dpf(pre_r, pre_c) + 1);
            }
        }

        if(isStart){
            return 1;
        }

        dp[r][c] = ret;

        return ret;
    }
}

File: 240126/정수 사각형 최장 증가 수열/lis-on-the-integer-grid.py
n = int(input())

import sys
sys.setrecursionlimit(n**2)


board = [list(map(int, input().split())) for _ in range(n)]
dp = [[-1 for _ in range(n)] for _ in range(n)]

inf = int(1e8)

dr = [-1, 1, 0, 0]
dc = [0, 0, -1, 1]

def in_range(r, c):
    return 0 <= r < n and 0 <= c < n

# dpf(r, c): 임의의 출발점에 대해서 도착점(r, c)까지 올 수 있는 최대거리
# dpf(r, c) = max(dpf(br, bc) where br, bc is adj && board[br][bc] < board[r][c]) + 1 <- 항상 식부터!!

def dpf(r, c):
    # dp 있다! 종료!
    if dp[r][c] != -1:
        return dp[r][c]

    # 점화식
    is_start = True
    ret = -inf
    for k in range(4):
        br, bc = r + dr[k], c + dc[k]
        if in_range(br, bc) and board[r][c] > board[br][bc]:
            is_start = False
            ret = max(ret, dpf(br, bc) + 1)  


    # 시작점 - 출발점일 때...
    if is_start:
        return 1

    dp[r][c] = ret

    return ret

max_ret = 0
for r in range(n):
    for c in range(n):
        max_ret = max(max_ret, dpf(r, c))

print(max_ret)